change: [M3-9881] - Improve behavior of VLANSelect component when creating a new VLAN

coliu-akamai opened the pull request:
## Description 📝
Updates UX for VLAN Select to be more user friendly

## Changes  🔄
- clears VlanSelect textfield if no vlan explicitly selected to make it clear that you must select a VLAN
- Adds region filtering for VLAN Select in Add Interface form
- Can change VLAN value via typing if already existing VLAN gets selected

## Preview 📷
no visual changes
Main behavior change:

| Before  | After   |
| ------- | ------- |
| <video src="https://github.com/user-attachments/assets/c2333803-9317-4612-980b-9d2e4eff2cb1" /> | <video src="https://github.com/user-attachments/assets/f3398bd3-1003-46a9-be20-d1a1ef21acbe" /> |

## How to test 🧪
have:
- Linode Interfaces flag + account capability
- two vlans in different regions

### Verification steps

- Confirm no regressions/changes in behavior besides these in VLANSelect on Create (both old and new UIs) and AddInterfaceForm
- Confirm VLANSelect textfield clears if no vlan option or 'create new vlan option' explicitly selected
- confirm VLANSelect in AddInterfaceForm only filters for the vlans in given region
- Confirm that if you've already selected a VLAN, and you use the keyboard to try to change your option (ex: backspacing a bunch), the select updates accordingly

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

👀 Doing a self review
❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
🤏 Splitting feature into small PRs
➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
🧪 Providing/improving test coverage
 🔐 Removing all sensitive information from the code and PR description
🚩 Using a feature flag to protect the release
👣 Providing comprehensive reproduction steps
📑 Providing or updating our documentation
🕛 Scheduling a pair reviewing session
📱 Providing mobile support
♿  Providing accessibility support

<br/>

- [ ] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [ ] All unit tests are passing
- [ ] TypeScript compilation succeeded without errors
- [ ] Code passes all linting rules

</details>

coliu-akamai commented on packages/manager/src/components/VLANSelect.tsx:
```diff
@@ -108,27 +111,24 @@ export const VLANSelect = (props: Props) => {
-      inputValue={selectedVLAN ? selectedVLAN.label : inputValue}
-      isOptionEqualToValue={(option1, options2) =>
-        option1.label === options2.label
+      inputValue={selectedVLAN && !open ? selectedVLAN.label : inputValue}
```
This + the useEffect addition should hopefully fix our ability to type in the select after choosing an existing vlan:

| Before  | After   |
| ------- | ------- |
| <video src="https://github.com/user-attachments/assets/7455bee7-9bc4-4449-a3c6-62792e6ff6ad" /> | <video src="https://github.com/user-attachments/assets/3c4460b0-8462-4865-a0e8-bb5f74abee7b" /> |

bnussman-akamai commented:
https://github.com/user-attachments/assets/6ae02ba4-26fa-465e-8be5-2e4eacfbecf9

I think users want the new VLAN to be selected onBlur. I think that's what's been tripping them up. 

I don't mind the current behavior personally, but I think users get confused when they type in a VLAN name but don't select the "Create" option in the dropdown. I

bnussman-akamai commented:
Seeing a small difference between this PR and prod. 

Prod seems to not filter options when one is selected but this PR does

| Prod  | This PR  |
|---|---|
| <video src="https://github.com/user-attachments/assets/b13cce03-62d9-4b63-b8cd-a6d40c1dde9f" />  |  <video src="https://github.com/user-attachments/assets/3660c24e-526f-482d-9240-5d138efb4ab5" />  |

hana-akamai commented on packages/manager/src/components/VLANSelect.tsx:
```diff
@@ -108,27 +111,31 @@ export const VLANSelect = (props: Props) => {
-      onBlur={onBlur}
+      onBlur={() => {
+        if (inputValue !== value) {
+          if (vlans.length === 1 && onChange) {
```
I think we should just stick with one behavior (clear on blur or keep value) instead of conditional behavior

coliu-akamai commented:
@bnussman-akamai @hana-akamai think what I pushed up fixed filtering + switched to only keeping value of what's typed inside the select!

unfortunately I had to remove the changes that addressed the issues [here](https://github.com/linode/manager/pull/12380#discussion_r2145858338) to fix filtering - will continue looking into how to fix typing in the select once we've chosen a value without messing up filtering, but not yet sure if there's a solution to keep both features 

current behavior:
<video src="https://github.com/user-attachments/assets/a481127a-7b81-4648-ad6a-ea04a1ec7002" />

coliu-akamai commented:
hmm I feel like I wouldn't expect clicking an option in a select to have different behavior than the other options / behavior besides just selecting. having a select with two different functionalities might be confusing

However since we provide the option to create a VPC/firewall within the Create Page (albeit as a separate action/button), matching that behavior here could be nice. (In the Add Interface drawer though, we don't provide the option to create a new VPC)

@bnussman-akamai what are your thoughts?

Should we get UX input?

bnussman-akamai reviewed (approved):
VLAN select feels great now 🎉
